File: src/android/src/br/com/kaiquecruz/udpconnection/UdpConnection.java
/*
 * Copyright 2018 Kaique Cruz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Based on Mustafa Alparslan's code from the blog: http://msdalp.github.io/2014/03/09/Udp-on-Android/
 */
package br.com.kaiquecruz.udpconnection;

import org.apache.cordova.*;
import java.lang.reflect.Field;
import java.lang.reflect.Method;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;
import java.net.UnknownHostException;
import android.content.Context;
import android.util.Log;

public class UdpConnection extends CordovaPlugin {
	
    private static final String CLIENT_SEND_AND_LISTEN = "clientSendAndListen";
    private String ip_server = "255.255.255.255";
    private int port_server = 5000, port_client = 4999;
    private DatagramSocket udpSocket;	
    private CallbackContext callbackContext;
	private static final String TAG = "UDP_CONNECTION";
    
    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
    }
    
    @Override
    public boolean execute(String action, JSONArray data, CallbackContext callbackContext)
      throws JSONException {
        
        this.callbackContext = callbackContext;
        
        if(action.equals(CLIENT_SEND_AND_LISTEN)) {
            return this.clientSendAndListen(callbackContext, data);
        }
        else {
            callbackContext.error("Incorrect action parameter: " + action);
        }
        
        return false;
    }
    
    /**
     *    This method is used to send a text message to the server and get the response.
     *
     *    @param    callbackContext        A Cordova callback context
     *    @param    data                JSON Array, with [0] being SSID to remove
     *    @return   if true returns the server's message response, false if failed
     */
    private boolean clientSendAndListen(CallbackContext callbackContext, JSONArray data) {
        Log.d(TAG, "UdpConnection: clientSendAndListen entered.");
        boolean run = true;
		try {
			this.udpSocket = new DatagramSocket(this.port_client);
			InetAddress serverAddr = InetAddress.getByName(this.ip_server);
			byte[] buf = data.getString(0).getBytes();
			DatagramPacket packet = new DatagramPacket(buf, buf.length, serverAddr, this.port_server);
			this.udpSocket.send(packet);
			while (run) {
				try {
					byte[] message = new byte[10000];
					DatagramPacket packetII = new DatagramPacket(message,message.length);
					Log.i(TAG, "UDP client: about to wait to receive");
					this.udpSocket.setSoTimeout(10000);//10s
					this.udpSocket.receive(packetII);
					String text = new String(message, 0, packetII.getLength());
					
					Log.d(TAG, "Received text: "+ text);
					
					callbackContext.success(text);
					
					return true;
				} catch (IOException e) {
					Log.e(TAG, "Error: "+ e);
					run = false;
					this.udpSocket.close();
					
					callbackContext.error(e.getMessage());
					
					return false;
				}
			}
		} catch (SocketException e) {
			Log.e(TAG, "Error: "+ e);
			callbackContext.error(e.getMessage());
			return false;
		} catch (UnknownHostException e) {
			e.printStackTrace();
			Log.e(TAG, "Error: "+ e);
			callbackContext.error(e.getMessage());
			return false;
		} catch (IOException e) {
			e.printStackTrace();
			Log.e(TAG, "Error: "+ e);
			callbackContext.error(e.getMessage());
			return false;
        }
    }
	
}

File: src/ios/UdpConnection.m
/*
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 
 Based on Changxi Wu's code from the github: https://gist.github.com/chancyWu/8349411
 */

#import "UdpConnection.h"

@implementation UdpConnection

CDVInvokedUrlCommand* cmd;
CDVPluginResult* pluginResult;

// receive the message to send data
- (void) clientSendAndListen:(CDVInvokedUrlCommand*)command {
    cmd = command;
    
    int socketSD = socket(PF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (socketSD <= 0) {
        NSLog(@"Error: Could not open socket.");
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: Could not open socket."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    // set socket options enable broadcast
    int broadcastEnable = 1;
    long ret = setsockopt(socketSD, SOL_SOCKET, SO_BROADCAST, &broadcastEnable, sizeof(broadcastEnable));
    if (ret) {
        NSLog(@"Error: Could not open set socket to broadcast mode");
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: Could not open set socket to broadcast mode"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    // bind the port
    struct sockaddr_in sockaddr;
    memset(&sockaddr, 0, sizeof(sockaddr));
    
    sockaddr.sin_len = sizeof(sockaddr);
    sockaddr.sin_family = AF_INET;
    sockaddr.sin_port = htons(4999); //CLIENT PORT
    sockaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    
    int status = bind(socketSD, (struct sockaddr *)&sockaddr, sizeof(sockaddr));
    if (status == -1) {
        close(socketSD);
        NSLog(@"Error: listenForPackets - bind() failed.");
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: listenForPackets - bind() failed."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    
    // Configure the port and ip we want to send to
    struct sockaddr_in broadcastAddr;
    memset(&broadcastAddr, 0, sizeof(broadcastAddr));
    broadcastAddr.sin_family = AF_INET;
    inet_pton(AF_INET, "255.255.255.255", &broadcastAddr.sin_addr);
    broadcastAddr.sin_port = htons(5000); //SERVER PORT
    
    const char * request = ((NSString *)[command.arguments objectAtIndex:0]).cString;
    ret = sendto(socketSD, request, strlen(request), 0, (struct sockaddr*)&broadcastAddr, sizeof(broadcastAddr));
    if (ret < 0) {
        NSLog(@"Error: Could not open send broadcast.");
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: Could not open send broadcast."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    
    // set timeout to 3 seconds.
    struct timeval timeV;
    timeV.tv_sec = 3;
    timeV.tv_usec = 0;
    
    if (setsockopt(socketSD, SOL_SOCKET, SO_RCVTIMEO, &timeV, sizeof(timeV)) == -1) {
        NSLog(@"Error: listenForPackets - setsockopt failed");
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: listenForPackets - setsockopt failed"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    
    // receive
    struct sockaddr_in receiveSockaddr;
    socklen_t receiveSockaddrLen = sizeof(receiveSockaddr);
    
    size_t bufSize = 10000;
    void *buf = malloc(bufSize);
    ssize_t result = recvfrom(socketSD, buf, bufSize, 0, (struct sockaddr *)&receiveSockaddr, &receiveSockaddrLen);
    
    NSData *data = nil;
    
    NSLog(@"result: %zd", result);
    
    if (result > 0) {
        if ((size_t)result != bufSize) {
            buf = realloc(buf, result);
        }
        data = [NSData dataWithBytesNoCopy:buf length:result freeWhenDone:YES];
        
        char addrBuf[INET_ADDRSTRLEN];
        if (inet_ntop(AF_INET, &receiveSockaddr.sin_addr, addrBuf, (size_t)sizeof(addrBuf)) == NULL) {
            addrBuf[0] = '\0';
        }
        
        NSString *msg = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:msg];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
    } else {
        free(buf);
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error receiving UDP packet"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
    }
    
}


// receive the message to send data
- (void) clientSendAndListenCreateProjectOnIOS:(CDVInvokedUrlCommand*)command {
    cmd = command;
    
    int socketSD = socket(PF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (socketSD <= 0) {
        NSLog(@"Error: Could not open socket.");
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: Could not open socket."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    // set socket options enable broadcast
    int broadcastEnable = 1;
    long ret = setsockopt(socketSD, SOL_SOCKET, SO_BROADCAST, &broadcastEnable, sizeof(broadcastEnable));
    if (ret) {
        NSLog(@"Error: Could not open set socket to broadcast mode");
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: Could not open set socket to broadcast mode"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    // bind the port
    struct sockaddr_in sockaddr;
    memset(&sockaddr, 0, sizeof(sockaddr));
    
    sockaddr.sin_len = sizeof(sockaddr);
    sockaddr.sin_family = AF_INET;
    sockaddr.sin_port = htons(4999); //CLIENT PORT
    sockaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    
    int status = bind(socketSD, (struct sockaddr *)&sockaddr, sizeof(sockaddr));
    if (status == -1) {
        close(socketSD);
        NSLog(@"Error: listenForPackets - bind() failed.");
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: listenForPackets - bind() failed."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    
    // Configure the port and ip we want to send to
    struct sockaddr_in broadcastAddr;
    memset(&broadcastAddr, 0, sizeof(broadcastAddr));
    broadcastAddr.sin_family = AF_INET;
    inet_pton(AF_INET, "255.255.255.255", &broadcastAddr.sin_addr);
    broadcastAddr.sin_port = htons(5000); //SERVER PORT
    
    const char * request = ((NSString *)[command.arguments objectAtIndex:0]).cString;
    ret = sendto(socketSD, request, strlen(request), 0, (struct sockaddr*)&broadcastAddr, sizeof(broadcastAddr));
    if (ret < 0) {
        NSLog(@"Error: Could not open send broadcast.");
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: Could not open send broadcast."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    
    // set timeout to 10 seconds.
    struct timeval timeV;
    timeV.tv_sec = 10;
    timeV.tv_usec = 0;
    
    if (setsockopt(socketSD, SOL_SOCKET, SO_RCVTIMEO, &timeV, sizeof(timeV)) == -1) {
        NSLog(@"Error: listenForPackets - setsockopt failed");
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error: listenForPackets - setsockopt failed"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
        
        return;
    }
    
    
    // receive
    struct sockaddr_in receiveSockaddr;
    socklen_t receiveSockaddrLen = sizeof(receiveSockaddr);
    
    size_t bufSize = 10000;
    void *buf = malloc(bufSize);
    ssize_t result = recvfrom(socketSD, buf, bufSize, 0, (struct sockaddr *)&receiveSockaddr, &receiveSockaddrLen);
    
    NSData *data = nil;
    
    NSLog(@"result: %zd", result);
    
    if (result > 0) {
        if ((size_t)result != bufSize) {
            buf = realloc(buf, result);
        }
        data = [NSData dataWithBytesNoCopy:buf length:result freeWhenDone:YES];
        
        char addrBuf[INET_ADDRSTRLEN];
        if (inet_ntop(AF_INET, &receiveSockaddr.sin_addr, addrBuf, (size_t)sizeof(addrBuf)) == NULL) {
            addrBuf[0] = '\0';
        }
        
        NSString *msg = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:msg];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
    } else {
        free(buf);
        close(socketSD);
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error receiving UDP packet"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:cmd.callbackId];
    }
    
}

@end